src/parquet/file/metadata-internal.cc

```diff
@@ -0,0 +1,650 @@
+  return impl->num_row_groups();
+}
+
+int32_t FileMetaDataReader::version() const {
+  return impl->version();
+}
+
+const std::string& FileMetaDataReader::created_by() const {
+  return impl->created_by();
+}
+
+int FileMetaDataReader::num_schema_elements() const {
+  return impl->num_schema_elements();
+}
+
+// MetaDataWriters
+// Update row-group Columns
+class SchemaMetaDataWriter::InternalSchemaMetaDataWriter {
+ public:
+  InternalSchemaMetaDataWriter() {}
+  ~InternalSchemaMetaDataWriter() {}
+  // Schema metadata
+  void PhysicalType(Type::type val) { schema_.__set_type(ToThrift(val)); }
+  void type_length(int32_t val) { schema_.__set_type_length(val); }
+  void num_children(int32_t val) { schema_.__set_num_children(val); }
+  void scale(int32_t val) { schema_.__set_scale(val); }
+  void precision(int32_t val) { schema_.__set_precision(val); }
+  void field_id(int32_t val) { schema_.__set_field_id(val); }
+  void name(const std::string& val) { schema_.__set_name(val); }
+  void ConvertedType(LogicalType::type val) {
+    schema_.__set_converted_type(ToThrift(val));
+  }
+  void RepetitionType(Repetition::type val) {
+    schema_.__set_repetition_type(ToThrift(val));
+  }
+  const uint8_t* get_buffer() { return reinterpret_cast<const uint8_t*>(&schema_); }
+  size_t get_buffer_size() { return sizeof(schema_); }
+
+ private:
+  format::SchemaElement schema_;
+};
+
+std::unique_ptr<SchemaMetaDataWriter> SchemaMetaDataWriter::GetSchemaMetaDataWriter() {
+  return std::unique_ptr<SchemaMetaDataWriter>(new SchemaMetaDataWriter());
+}
+
+SchemaMetaDataWriter::SchemaMetaDataWriter()
+    : impl{std::unique_ptr<InternalSchemaMetaDataWriter>(
+          new InternalSchemaMetaDataWriter())} {}
+
+SchemaMetaDataWriter::~SchemaMetaDataWriter() {}
+
+const uint8_t* SchemaMetaDataWriter::get_buffer() {
+  return impl->get_buffer();
+}
+
+size_t SchemaMetaDataWriter::get_buffer_size() {
+  return impl->get_buffer_size();
+}
+
+// schema metadata
+void SchemaMetaDataWriter::PhysicalType(Type::type val) {
+  impl->PhysicalType(val);
+}
+
+void SchemaMetaDataWriter::type_length(int32_t val) {
+  impl->type_length(val);
+}
+
+void SchemaMetaDataWriter::num_children(int32_t val) {
+  impl->num_children(val);
+}
+
+void SchemaMetaDataWriter::scale(int32_t val) {
+  impl->scale(val);
+}
+
+void SchemaMetaDataWriter::precision(int32_t val) {
+  impl->precision(val);
+}
+
+void SchemaMetaDataWriter::field_id(int32_t val) {
+  impl->field_id(val);
+}
+
+void SchemaMetaDataWriter::name(const std::string& val) {
+  impl->name(val);
+}
+
+void SchemaMetaDataWriter::ConvertedType(LogicalType::type val) {
+  impl->ConvertedType(val);
+}
+
+void SchemaMetaDataWriter::RepetitionType(Repetition::type val) {
+  impl->RepetitionType(val);
+}
+
+// row-group metadata
+class RowGroupMetaDataWriter::InternalRowGroupMetaDataWriter {
+ public:
+  explicit InternalRowGroupMetaDataWriter(const WriterProperties& props)
+      : properties_(props), num_columns_(0) {}
```

Comment:
To keep the properties local to the writer seems like a better design to me. You don't need to manage the properties externally. Also protects from accidental modification of properties while writing the metadata.